src/threshold.jl

```diff
@@ -1,30 +1,41 @@
-function threshold_dithering(
-    img::AbstractMatrix{<:Gray}, threshold_map::AbstractMatrix{<:Gray}; to_linear=false
-)::BitMatrix
-    size(img) ≢ size(threshold_map) && throw(
-        DimensionMismatch("Arguments img and threshold_map need to have same dimensions"),
-    )
+abstract type AbstractThresholdDither <: AbstractDither end
 
+function dither(
+    img::AbstractMatrix{<:Gray}, alg::AbstractThresholdDither; to_linear=false
+)::Matrix{Gray{Bool}}
+    tmap = alg(img)
     to_linear && (img = mappedarray(srgb2linear, img))
-    return img .> threshold_map
+    return img .> tmap
```

Comment:
The return type annotation is not usually needed:

```suggestion
    return reinterpret(Gray{Bool}, img .> tmap)
```

and it's not always performant:

```julia
julia> function foo()::Matrix{Gray{Bool}}
           return rand(Bool, 64, 64)
       end
foo (generic function with 1 method)

julia> function bar()::Matrix{Gray{Bool}}
           return reinterpret(Gray{Bool}, rand(Bool, 64, 64))
       end
bar (generic function with 1 method)

julia> @btime foo();
  2.997 μs (2 allocations: 8.38 KiB)

julia> @btime bar();
  2.673 μs (2 allocations: 8.38 KiB)
```